blog/2024-10-28-resource-api/README.md

```diff
@@ -0,0 +1,386 @@
+      .then(res => res.json()) as Promise<Book[]>
+  }
+}
+```
+
+```ts
+// Komponente
+booksResource = resource({
+  loader: () => this.bs.getAll()
+});
+```
+
+
+### Variante 2: Angulars `HttpClient` und Observable
+
+Wir verwenden wir üblich den `HttpClient` von Angular, sodass die Methode `getAll()` ein Observable zurückgibt.
+Um den Loader zu definieren, müssen wir das Observable mithilfe von `firstValueFrom()` in eine Promise umwandeln.
+
+```ts
+@Injectable({ /* ... */ })
+export class BookStoreService {
+  // ...
+  getAll(): Observable<Book[]> {
+    return this.http.get<Book[]>(this.apiUrl + '/books');
+  }
+}
+```
+
+```ts
+// Komponente
+booksResource = resource({
+  loader: () => firstValueFrom(this.bs.getAll())
+});
+```
+
+
+## Daten lesen
+
+Der Loader wird sofort ausgeführt, sobald das Resource-Objekt initialisiert wird. Die Resource verarbeitet die Antwort und bietet folgende Signals an, um mit den Daten zu arbeiten:
+
+- `value`: geladene Daten, hier `Book[]`
+- `status`: Zustand der Resource vom Typ `ResourceStatus`, z. B. *Resolved* oder *Idle*, siehe nächster Abschnitt
+- `error`: Fehler
+
+Die geladenen Bücher können wir also wir folgt im Template anzeigen:
+
+```html
+{{ booksResource.value() | json }}
+
+@for(book of booksResource.value(); track book.isbn) {
+  <p>{{ book.title }}</p>
+}
+```
+
+## Status der Resource
+
+Mithilfe des Signals `status` können wir den Zustand der Resource auswerten, z. B. um einen Ladeindikator anzuzeigen. Alle Werte von `status` sind Felder aus dem [Enum `ResourceStatus`](https://next.angular.dev/api/core/ResourceStatus):
+
+| Status aus `ResourceStatus` | Beschreibung                                                                         |
+| --------------------------- | ------------------------------------------------------------------------------------ |
+| `Idle`                      | Es ist kein Request definiert und es wird nichts geladen. `value()` ist `undefined`. |
+| `Error`                     | Das Laden ist fehlgeschlagen. `value()` ist `undefined`.                             |
+| `Loading`                   | Die Resource lädt gerade.                                                            |
+| `Reloading`                 | Die Resource lädt gerade neu, nachdem das Neuladen angefordert wurde.                |
+| `Resolved`                  | Das Laden ist abgeschlossen.                                                         |
+| `Local`                     | Der Wert wurde lokal überschrieben.                                                  |
+
+Für einen Ladeindikator können wir z. B. so vorgehen und ein Computed Signal verwenden, das den Zustand auswertet und ein Boolean bereitstellt:
+
+```ts
+import { resource, computed, ResourceStatus } from '@angular/core';
+// ...
+
+isLoading = computed(() => {
+  return this.booksResource.status() === ResourceStatus.Loading ||
+    this.booksResource.status() === ResourceStatus.Reloading;
+  });
```

Comment:
```suggestion
  return [ResourceStatus.Loading, ResourceStatus.Reloading]
    .includes(this.booksResource.status())
  });
```

Ich fände es etwas schöner weil nicht 2x `this.booksResource.status()`. Aber auch hier hänge ich nicht am Vorschlag.

Reply:
Ich würde es ab 3 Zuständen bestimmt auch mit `includes()` machen, aber um schnell die Idee zu formulieren, fand ich den direkten Vergleich besser verständlich.
@JohannesHoppe darf entscheiden ;-)

Reply:
Für das Beispiel reicht es auch aus, wenn wir nur `Loading` nehmen, den Rest kann man sich dann selbst denken.

Reply:
Wäre auch okay!

Reply:
Würde auch nicht funktionieren, weil das Enum Zahlen ausgibt. 😇